### PR TITLE
[libc] fix type generic stdc_leading_zeros for GCC

### PR DESCRIPTION
**Pull request by nickdesaulniers**
GCC does not support _Generic in C++ mode. Use inline function definitions
with function overloading in __cplusplus mode.

### CODE REVIEW

File: libc/test/include/CMakeLists.txt [hunk 1]
```diff
@@ -27,13 +27,5 @@ if (LLVM_LIBC_FULL_BUILD)
     DEPENDS
       libc.include.llvm-libc-macros.stdbit_macros
       libc.include.stdbit
-      # Intentionally do not depend on libc.src.stdbit.*. The include test is
-      # simply testing the macros provided by stdbit.h, not the implementation
-      # of the underlying functions which the type generic macros may dispatch
-      # to.
```
Comment (nickdesaulniers) [inline libc/test/include/CMakeLists.txt:33]: Ah, I should probably keep this part

Comment (nickdesaulniers) [reply libc/test/include/CMakeLists.txt]: done in 42c3fb895513

**Review by michaelrj-google — APPROVED** [review]

**Comment by jplehr** [comment]
I think this broke the AMDGPU libc bot. https://lab.llvm.org/staging/#/builders/11/builds/1282
@jhuber6 can you take a look?

**Comment by nickdesaulniers** [comment]
Fixed in https://github.com/llvm/llvm-project/commit/0129ff17a9885d76d3700fee56500848c28e3fc1. Sorry about that!

@jhuber6 I think we can simplify the condition further to just:

```cmake
if(libc.include.stdbit IN_LIST TARGET_PUBLIC_HEADERS)
```
(I'm guessing that `TARGET_PUBLIC_HEADERS` is empty when `LLVM_LIBC_FULL_BUILD` is not set?)

**Comment by nickdesaulniers** [comment]
Also, we can probably add these to the gpu headers file.

**Comment by jhuber6** [comment]
> Also, we can probably add this macro to the gpu headers file.

It's possible, I think I tried and it had some other problem so I just ignored it for now. We can reevaluate later.

**Comment by jhuber6** [comment]
> Fixed in [0129ff1](https://github.com/llvm/llvm-project/commit/0129ff17a9885d76d3700fee56500848c28e3fc1). Sorry about that!
> 
> @jhuber6 I think we can simplify the condition further to just:
> 
> ```cmake
> if(libc.include.stdbit IN_LIST TARGET_PUBLIC_HEADERS)
> ```
> 
> (I'm guessing that `TARGET_PUBLIC_HEADERS` is empty when `LLVM_LIBC_FULL_BUILD` is not set?)

Don't think so, but we could alter the logic to do that if it's the desire behavior.
```diff
diff --git a/libc/CMakeLists.txt b/libc/CMakeLists.txt
index 6d385849b6a6..66d8f702d420 100644
--- a/libc/CMakeLists.txt
+++ b/libc/CMakeLists.txt
@@ -314,7 +314,7 @@ else()
 endif()
 
 # Check headers.txt
-if(EXISTS "${LIBC_CONFIG_PATH}/headers.txt")
+if(EXISTS "${LIBC_CONFIG_PATH}/headers.txt" AND LLVM_LIBC_FULL_BUILD)
     include("${LIBC_CONFIG_PATH}/headers.txt")
 elseif(LLVM_LIBC_FULL_BUILD)
   message(FATAL_ERROR "${LIBC_CONFIG_PATH}/headers.txt file not found and fullbuild requested.")
```
Or something. might break something else, who knows.

**Pull request merged**